### PR TITLE
definition of Scala data structures for representing trigger logging metric data

### PR DESCRIPTION
**Pull request by carlpulley-da**
- definition of Scala data structures for representing trigger logging metric data
- addition of a singleton factory for creating instances of simulation library
- some simulation library refactors

### CODE REVIEW

File: triggers/tests/src/test/scala/com/digitalasset/daml/lf/engine/trigger/TriggerRuleSimulationLib.scala [hunk 1]
```diff
@@ -6,46 +6,106 @@ package com.daml.lf.engine.trigger
...
-import com.daml.logging.LoggingContextOf
+import com.daml.lf.speedy.SValue.SList
+import com.daml.lf.speedy.SValue
+import com.daml.platform.services.time.TimeProviderType
 import com.daml.util.Ctx
 
 import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+
+object TriggerRuleMetrics {
```
Comment (carlpulley-da) [inline triggers/tests/src/test/scala/com/digitalasset/daml/lf/engine/trigger/TriggerRuleSimulationLib.scala:27]: High level view of logging metric data the simulation library will be using

File: triggers/tests/src/test/scala/com/digitalasset/daml/lf/engine/trigger/TriggerRuleSimulationLib.scala [hunk 2]
```diff
@@ -6,46 +6,106 @@ package com.daml.lf.engine.trigger
...
   def initialStateLambda(
       acs: Seq[CreatedEvent]
   ): Future[(Seq[TriggerContext[SubmitRequest]], SValue)] = {
-    val clientTime: Timestamp =
-      Timestamp.assertFromInstant(
-        Runner.getTimeProvider(RunnerConfig.DefaultTimeProviderType).getCurrentTime
-      )
-    val killSwitch = KillSwitches.shared("init-state-simulation")
-    val initStateGraph = GraphDSL.createGraph(Sink.last[SValue]) { implicit gb => saveLastState =>
+    def initStateGraph = GraphDSL.createGraph(Sink.last[SValue]) { implicit gb => saveLastState =>
```
Comment (carlpulley-da) [inline triggers/tests/src/test/scala/com/digitalasset/daml/lf/engine/trigger/TriggerRuleSimulationLib.scala:101]: As the `initState` lambdas will run twice, we need to use def's here

File: triggers/tests/src/test/scala/com/digitalasset/daml/lf/engine/trigger/TriggerRuleSimulationLib.scala [hunk 3]
```diff
@@ -66,13 +127,15 @@ class TriggerRuleSimulationLib[UserState](runner: Runner) {
       state: SValue,
       message: SValue,
   ): Future[(Seq[TriggerContext[SubmitRequest]], SValue)] = {
-    val lambdaKillSwitch = KillSwitches.shared("update-state-simulation")
-    val updateStateGraph = GraphDSL.createGraph(Sink.last[SValue]) { implicit gb => saveLastState =>
+    def updateStateGraph = GraphDSL.createGraph(Sink.last[SValue]) { implicit gb => saveLastState =>
```
Comment (carlpulley-da) [inline triggers/tests/src/test/scala/com/digitalasset/daml/lf/engine/trigger/TriggerRuleSimulationLib.scala:130]: As the `updateState` lambdas will run twice we need defs here

File: triggers/tests/src/test/scala/com/digitalasset/daml/lf/engine/trigger/TriggerRuleSimulationLib.scala [hunk 4]
```diff
@@ -6,93 +6,210 @@ package com.daml.lf.engine.trigger
...
+    triggerId: Identifier,
+    triggerConfig: TriggerRunnerConfig,
+    client: LedgerClient,
+    timeProviderType: TimeProviderType,
+    applicationId: ApplicationId,
+    triggerParties: TriggerParties,
+) {
+
+  private[this] val logObserver: (String, TriggerLogContext) => Unit = { (_, _) =>
+    // TODO: a future PR will provide a more meaningful implementation
```
Comment (carlpulley-da) [inline triggers/tests/src/test/scala/com/digitalasset/daml/lf/engine/trigger/TriggerRuleSimulationLib.scala:78]: Once the structured logging to trigger metric case class translation is defined, we can hook calls to the translation function here

**Review by remyhaemmerle-da — APPROVED** [review]

**Pull request merged**